### PR TITLE
Check if management server (actuator) is running on different port while constructing skip patterns

### PR DESCRIPTION
**Pull request by reta**
The Issue
======

While `Spring Cloud Sleuth` is constructing the skip patterns, the management/actuator web endpoints should be excluded from traces (addressed in https://github.com/spring-cloud/spring-cloud-sleuth/pull/1150/files and related PRs). However, when the management/actuator server is running on a different (from the application server) port, the skip patterns are still constructed using the application server context path.

Example:
```
server.port = 8080
server.servlet.contextPath = /api
management.server.port = 9080
management.endpoints.web.base-path = /mgt
```

The expected skip patterns (for `/info` and `health|` endpoints): 
```
/mgt/(info|info/.*|health|health/.*)
```

The current skip patterns (for `/info` and `health|` endpoints) are still using application context path: 
```
/api/mgt/(info|info/.*|health|health/.*)
```

Another  issue shows up when the management/actuator web endpoints are bound to root context '/', for example:

```
server.port = 8080
server.servlet.contextPath = /api
management.server.port = 9080
management.endpoints.web.base-path = /
```

In this case, the some application resources are going to be ignored by `SleuthHttpSampler` since the generated skip patterns will include `/api/(info|info/.*|health|health/.*)`

Suggested Solution
======
Take the management/actuator server port configuration into account and adjust the skip patterns accordingly.

Thank you!

PS: Not sure if this is still required, but the contributor agreement has been signed.

### CODE REVIEW

**Comment by shakuzen** [comment]
Multiple skip patterns are combined to get the final full pattern. Management properties are considered in a separate pattern.

Currently the case of the management port being different is covered by the following code, which would contribute a pattern `/mgt.*` with the example properties given. The other configuration class in the same file will also contribute a needless (wrong) pattern, but the endpoints should be skipped by the former pattern nonetheless.

https://github.com/spring-cloud/spring-cloud-sleuth/blob/299e21e7ecad69fce42365f84dc653f457f48e54/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebAutoConfiguration.java#L74-L102

I added a test that shows how it all gets put together:

```java
@Test
public void gh1208() {
	ServerProperties serverProperties = new ServerProperties();
	serverProperties.setPort(8080);
	serverProperties.getServlet().setContextPath("/api");
	WebEndpointProperties webEndpointProperties = new WebEndpointProperties();
	ManagementServerProperties managementServerProperties = new ManagementServerProperties();
	managementServerProperties.setPort(9080);
	managementServerProperties.getServlet().setContextPath("/mgt");
	EndpointsSupplier<ExposableWebEndpoint> endpointsSupplier = () -> {
		ExposableWebEndpoint infoEndpoint = createEndpoint("info");
		ExposableWebEndpoint healthEndpoint = createEndpoint("health");

		return Arrays.asList(infoEndpoint, healthEndpoint);
	};

	TraceWebAutoConfiguration traceWebAutoConfiguration = new TraceWebAutoConfiguration();
	traceWebAutoConfiguration.patterns = Arrays.asList(
			() -> TraceWebAutoConfiguration.ManagementSkipPatternProviderConfig
					.getPatternForManagementServerProperties(
							managementServerProperties),
			() -> TraceWebAutoConfiguration.ActuatorSkipPatternProviderConfig
					.getEndpointsPatterns(serverProperties, webEndpointProperties,
							endpointsSupplier));
	Pattern pattern = traceWebAutoConfiguration.sleuthSkipPatternProvider()
			.skipPattern();
	System.out.println(pattern); // /mgt.*|/api/actuator/(info|info/.*|health|health/.*)
	assertThat(pattern.pattern()).contains("/mgt.*");
}
```

Minimally, the pattern should just be `/mgt.*` or perhaps more precisely `/mgt|/mgt/.*`, but functionally, it should be working now as far as the endpoints being skipped.

**Comment by reta** [comment]
Thank @shakuzen, this is exactly where the issue is coming from, so the configuration which made me wondering why all our production endpoints are not being traced is this:

- application port on 8080, path `/api`
- management port on 9080, path `/`

```
server.port = 8080
server.servlet.contextPath = /api
management.server.port = 9080
management.servlet.contextPath = /
```

The example you provided will cause to all production endpoints being skipped since the management endpoint is bound to root context (the issue is present in `2.1.0.RELEASE` but code in master has changed slightly).

**Comment by marcingrzejszczak** [comment]
@shakuzen do you have sth to add?

**Comment by reta** [comment]
Thanks for checking @marcingrzejszczak, I updated the description to have more details. The issue turned to be a bit more complex (luckily the problematic `ServerSkipPatternProviderConfig` class is gone in master, it was causing the major issue by completely shading the application context). But the PR is still valid (I think), since depending on the management / actuator server configuration, the management/actuator web endpoints are going to be hosted under different context paths (and have to be excluded from traces).

**Comment by shakuzen** [comment]
As a more general comment, reading through the tests (thank you for adding more!) I felt like we should convert most of the tests to using Spring Boot's `ApplicationContextRunner`. We'll more easily be able to test the outcome of the whole auto-configuration class, as opposed to the unit test focused approach now. For catching the kind of issues we've been seeing in this area, I think testing the whole auto-configuration is a better approach and would be more clear. That said, I think we can do that outside of this pull request.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebAutoConfiguration.java [hunk 1]
```diff
@@ -142,14 +141,23 @@ private static String getPathPrefix(String contextPath, String actuatorBasePath)
 		}
 
 		@Bean
-		public SingleSkipPattern skipPatternForActuatorEndpoints(
+		@ConditionalOnManagementPort(ManagementPortType.SAME)
```
[the remark this answers is not in the document]
Comment (reta) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebAutoConfiguration.java]: Yeah, me too, discovered them while dug deeper into the management bootstrapping process :)

**Comment by reta** [comment]
Thanks a lot for the review and suggestion, @shakuzen , it make sense to me and and has been incorporated into the PR (indeed, the relationship between application context part, management context path and management web endpoints context path is tricky).

Regarding your suggestion to refactor test cases, absolutely agree. If you don't mind, I could submit another PR to switch them off to the context-based ones so they are going to be run against real application configuration (should be just a bit slower comparing to JUnit test cases).

Thank you.

**Comment by shakuzen** [comment]
> Regarding your suggestion to refactor test cases, absolutely agree. If you don't mind, I could submit another PR to switch them off to the context-based ones so they are going to be run against real application configuration (should be just a bit slower comparing to JUnit test cases).

I'm sure @marcingrzejszczak would be happy to get a PR for it, but I'll let him speak for himself. I'll be happy to review.

**Comment by marcingrzejszczak** [comment]
Great work @reta , thanks for the review @shakuzen

**Comment by marcingrzejszczak** [comment]
> I'm sure @marcingrzejszczak would be happy to get a PR for it, but I'll let him speak for himself. I'll be happy to review.

@shakuzen is reading my mind ;)

**Comment by reta** [comment]
Thanks @shakuzen and @marcingrzejszczak , will help out with test refact shorly :)

**Pull request merged**